warehouse/accounts/views.py

```diff
@@ -735,6 +744,104 @@ def _get_two_factor_data(request, _redirect_to="/"):
     return two_factor_data
 
 
+@view_config(
+    route_name="accounts.verify-project-role",
+    uses_session=True,
+    permission="manage:user",
+    has_translations=True,
+)
+def verify_project_role(request):
+    token_service = request.find_service(ITokenService, name="email")
+    user_service = request.find_service(IUserService, context=None)
+
+    def _error(message):
+        request.session.flash(message, queue="error")
+        return HTTPSeeOther(request.route_path("manage.account"))
+
+    try:
+        token = request.params.get("token")
+        data = token_service.loads(token)
+    except TokenExpired:
+        return _error(request._("Expired token: request a new project role invite"))
+    except TokenInvalid:
+        return _error(request._("Invalid token: request a new project role invite"))
+    except TokenMissing:
+        return _error(request._("Invalid token: no token supplied"))
+
+    # Check whether this token is being used correctly
+    if data.get("action") != "email-project-role-verify":
+        return _error(request._("Invalid token: not an email verification token"))
+
+    user = user_service.get_user(data.get("user_id"))
+    if user != request.user:
+        return _error(request._("Invite is not valid"))
+
+    project = (
+        request.db.query(Project).filter(Project.id == data.get("project_id")).one()
+    )
+    desired_role = data.get("desired_role")
+
+    role_invite = (
+        request.db.query(RoleInvitation)
+        .filter(RoleInvitation.project == project)
+        .filter(RoleInvitation.user == user)
+        .one()
+    )
+
+    if role_invite.invite_status != RoleInvitationStatus.Pending.value:
+        return _error(
+            request._("Role invitation is already assigned or has been revoked.")
+        )
```

Comment:
Maybe we can provide some more specific errors here based on what `invite_status` is instead of what it is not, as well as a generic fallback error message if it falls through all the conditionals?